// chatexchange/dist/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseAgoString = exports.arrayToKvp = exports.lazy = exports.delay = void 0;
const ChatExchangeError_1 = __importDefault(require("./Exceptions/ChatExchangeError"));
/**
 * @module Utils
 */
/**
 * Helper function to resolve promise after ms.
 *
 * @function
 * @param {number} ms Number of milliseconds to delay
 * @returns {Promise<void>} A promise that resovles after ms milliseconds
 */
const delay = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
exports.delay = delay;
/**
 * Helper function to provide promises from getters/setters.
 * Used to lazily initialize values when the getter returns undefined.
 *
 * @function
 * @param {() => any} getter A function to return the value
 * @param {() => Promise<void>} updater A function that sets a value, that can be later retrieved from the getter
 * @returns {Promise<T>} A promise that returns the value that was set from the updater parameter
 */
async function lazy(getter, updater) {
    let result = getter();
    if (typeof result !== "undefined") {
        return result;
    }
    await updater();
    result = getter();
    if (typeof result === "undefined") {
        throw new ChatExchangeError_1.default("Unable to find field.");
    }
    return result;
}
exports.lazy = lazy;
/**
 * Helper function to convert an array, to key/value pairs.
 * ie. <code>['foo', 'bar', 'meow', 'rawr']</code> => <code>{foo: 'bar', meow: 'rawr'}</code>
 *
 * @function
 * @param {Array<{ toString(): string }>} array The array to convert
 * @returns {object} The object that was converted
 */
const arrayToKvp = (array) => {
    const kvp = {};
    array.forEach((val, idx) => {
        if (idx % 2 === 1) {
            const key = array[idx - 1];
            kvp[key.toString()] = val;
        }
    });
    return kvp;
};
exports.arrayToKvp = arrayToKvp;
// tslint:disable:object-literal-sort-keys
const suffixes = {
    s: 1,
    m: 60,
    h: 3600,
    d: 86400,
    y: 31536000,
};
// tslint:enable:object-literal-sort-keys
/**
 * Helper function to parse time strings (Mainly on the profile pages) into seconds.<br />
 * For example: <code>2m ago</code> => <code>120</code>
 *
 * @function
 * @param {string} text The string of text to parse. ie <code>5s ago</code>
 * @throws {ChatExchangeError} If the string doesn't match the format suffix (s/m/h/d/y).
 * @returns {number} The number
 */
const parseAgoString = (value) => {
    const text = value.toString();
    const valMap = {
        "n/a": -1,
        "just now": 0,
    };
    const val = valMap[text];
    if (val !== void 0)
        return val;
    const [str] = text.split(" ");
    const char = str.slice(-1);
    if (suffixes[char] === void 0) {
        throw new ChatExchangeError_1.default(`Suffix Character Unrecognized: ${char}`);
    }
    const time = parseInt(str.slice(0, -1), 10);
    return time * suffixes[char];
};
exports.parseAgoString = parseAgoString;
