// chatexchange/dist/WebsocketEvent.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebsocketEvent = exports.ChatEventType = void 0;
const Message_1 = __importDefault(require("./Message"));
/**
 * @summary chat event type options
 * @see https://meta.stackexchange.com/a/218443/786798
 */
var ChatEventType;
(function (ChatEventType) {
    ChatEventType[ChatEventType["MESSAGE_POSTED"] = 1] = "MESSAGE_POSTED";
    ChatEventType[ChatEventType["MESSAGE_EDITED"] = 2] = "MESSAGE_EDITED";
    ChatEventType[ChatEventType["USER_JOINED"] = 3] = "USER_JOINED";
    ChatEventType[ChatEventType["USER_LEFT"] = 4] = "USER_LEFT";
    ChatEventType[ChatEventType["ROOM_RENAMED"] = 5] = "ROOM_RENAMED";
    ChatEventType[ChatEventType["STARS_CHANGED"] = 6] = "STARS_CHANGED";
    ChatEventType[ChatEventType["USER_MENTIONED"] = 8] = "USER_MENTIONED";
    ChatEventType[ChatEventType["MESSAGE_FLAGGED"] = 9] = "MESSAGE_FLAGGED";
    ChatEventType[ChatEventType["MESSAGE_DELETED"] = 10] = "MESSAGE_DELETED";
    ChatEventType[ChatEventType["FILE_ADDED"] = 11] = "FILE_ADDED";
    ChatEventType[ChatEventType["MESSAGE_REPLIED"] = 18] = "MESSAGE_REPLIED";
    ChatEventType[ChatEventType["MESSAGE_MOVED_OUT"] = 19] = "MESSAGE_MOVED_OUT";
    ChatEventType[ChatEventType["MESSAGE_MOVED_IN"] = 20] = "MESSAGE_MOVED_IN";
})(ChatEventType = exports.ChatEventType || (exports.ChatEventType = {}));
class WebsocketEvent extends Message_1.default {
    /**
     * @param {Client} client main chatexchange Client class instance
     * @param {ChatEvent} websocketMsg message from the chat websocket
     */
    constructor(client, websocketMsg) {
        const { message_id, content, room_id, room_name, event_type, time_stamp, target_user_id, user_id, user_name, } = websocketMsg;
        super(client, message_id, {
            content,
            roomId: room_id,
            roomName: room_name,
        });
        this.eventType = event_type;
        this.timeStamp = time_stamp;
        this.targetUserId = target_user_id;
        this.userId = user_id;
        this.userName = user_name;
    }
}
exports.WebsocketEvent = WebsocketEvent;
exports.default = WebsocketEvent;
